sys/parsoid.js

```diff
@@ -125,109 +125,267 @@ class ParsoidService {
-        newReq.uri = `${this.parsoidHost}/${domain}/v3/${path}/pagebundle/` +
-            `${encodeURIComponent(rp.title)}/${rp.revision}`;
-        return hyper.request(newReq);
-    }
-
-    saveParsoidResultToLatest(hyper, req, tid, parsoidResp) {
+    _saveParsoidResultToFallback(hyper, req, parsoidResp) {
         const rp = req.params;
+        const dataParsoidResponse = parsoidResp.body['data-parsoid'];
+        const htmlResponse = parsoidResp.body.html;
+        const tid = mwUtil.parseETag(htmlResponse.headers.etag).tid;
         return hyper.put({
-            uri: this.getOldLatestBucketURI(rp, 'all', tid),
-            body: {
-                html: parsoidResp.body.html,
-                'data-parsoid': parsoidResp.body['data-parsoid']
-            }
+            uri: this.getStashBucketURI(rp.domain, rp.title, rp.revision, tid),
+            // Note. The headers we are storing here are for the whole pagebundle response.
+            // The individual components of the pagebundle contain their own headers that
+            // which are used to generate actual responses.
+            headers: {
+                'x-store-etag': htmlResponse.headers.etag,
+                'content-type': 'application/octet-stream',
+                'x-store-content-type': 'application/json'
+            },
+            body: Buffer.from(JSON.stringify({
+                'data-parsoid': dataParsoidResponse,
+                html: htmlResponse
+            }))
         });
     }
 
-    saveParsoidResultToFallback(hyper, req, tid, parsoidResp) {
-        const rp = req.params;
+    /**
+     * Saves the Parsoid pagebundle result to the latest bucket.
+     * // TODO: Optimization opportunity. We look what's in the
+     * // latest bucket yet again and make the store request a no-op if
+     * // it's not really the latest content, but we have already looked into
+     * // the latest bucket, thus we can somehow pass the data over here
+     * // so that we don't do several checks.
+     * @param {HyperSwitch} hyper the hyper object for request routing
+     * @param {string} domain the domain name
+     * @param {string} title the page title
+     * @param {Object} parsoidResp the response received from Parsoid.
+     * @return {Promise<Object>}
+     */
+    saveParsoidResultToLatest(hyper, domain, title, parsoidResp) {
         const dataParsoidResponse = parsoidResp.body['data-parsoid'];
         const htmlResponse = parsoidResp.body.html;
-        return hyper.put({
-            uri: this.getStashBucketURI(rp.domain, rp.title, rp.revision, tid),
+        return hyper.get({ uri: this.getLatestBucketURI(domain, title) })
+        .then((existingRes) => {
+            const existingRev = mwUtil.parseETag(existingRes.headers.etag).rev;
+            const newRev = mwUtil.parseETag(parsoidResp.headers.etag).rev;
+            if (Number.parseInt(newRev, 10) >= Number.parseInt(existingRev, 10)) {
+                throw new HTTPError({ status: 404 });
```

Comment:
Using 404 is confusing to me. Perhaps 412 or similar would be better? Or a custom code, even better.

Also, this compares revisions. If they are the same, the code seems to assume that the content provided by Parsoid is newer than the one in storage, which in 99.9% of the cases will be true, but due to the async nature of updates and multiple DCs, there will be corner cases where the update will not happen. I think for this time around that is ok, but let's at least leave a comment to that effect.